### ISSUE TITLE
stdarch - compile-time errors on invalid arguments

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#85264]
As noted in https://github.com/rust-lang/rust/issues/85155 which tracks the diagnostic issue, the standard library previously accepted code (with differing behavior) on a number of intrinsics with "intended to be const" arguments. Since https://github.com/rust-lang/rust/pull/83278, though, this code is rejected at compile-time.

For example, this [code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d7828dc84aed53820bd4fa24270a71d7) compiles on stable but not 1.54 (nightly as of this writing). I think at minimum we need to identify the set of code that changed behaviors and document that in release notes.

```rust
#[cfg(target_arch = "x86_64")]
fn main() {
    use std::arch::x86_64::*;
    unsafe {
        let a: __m128 = _mm_setzero_ps();
        let b: __m128 = _mm_setzero_ps();
        let _blended = _mm_blend_ps(a, b, 0x33);
    }
}
```

**Comment by Mark-Simulacrum** [comment]
cc @Amanieu -- on the point about documenting this change in behavior in the release notes, would you be up for either identifying the set of intrinsics which changed behavior or documenting how to do so?

**Comment by Amanieu** [comment]
Looking back at the crater run, we only ran it in `check` mode, which doesn't perform monomorphization. This is the reason why this wasn't caught earlier.

The full crater run for beta is currently running, so we should be able to get a clearer picture of the breakage once it is done.

Documenting this is difficult: the new behavior affects all intrinsics which use const arguments. A quick grep of stdarch shows 1741 intrinsics, of which about ~150 are stable.

**Comment by Mark-Simulacrum** [comment]
This is in 1.54, not 1.53, so the currently running beta run won't catch anything, right? It may make sense to run a crater on the PR itself, though.

It is worrying that so many intrinsics are affected towards a compile-error, vs. e.g. having a warning - perhaps deny-by-default - like clippy's on the atomic Ordering params...

**Comment by Amanieu** [comment]
@craterbot run mode=build-only end=881c1ac408d93bb7adaa3a51dabab9266e82eee8 start=ff34b919075f35a1787659e9c448a34b06bab8de name=issue-85264

**Comment by lqd** [comment]
The mechanism used to validate the immediates using const generics is:
- [this `ValidateConstImm<const IMM: i32, const MIN: i32, const MAX: i32>` struct](https://github.com/rust-lang/stdarch/blob/c14e98417feb406df66e821ccd81e1293b4baa6f/crates/core_arch/src/macros.rs#L3-L10) offering the validation feature: a post-monomorphization error while evaluating the `const VALID: () = { let _ = 1 / ((IMM >= MIN && IMM <= MAX) as usize); }`
- a [set of `static_assert_` macros](https://github.com/rust-lang/stdarch/blob/c14e98417feb406df66e821ccd81e1293b4baa6f/crates/core_arch/src/macros.rs#L12-L79) forcing the evaluation of the `VALID` const for the const generic parameters of a given intrinsic: e.g. the `_mm_blend_ps` above is [defined here](https://github.com/rust-lang/stdarch/blob/625dffe0fcbfd6a877f517b04ddb0b3c1177822f/crates/core_arch/src/x86/sse41.rs#L133-L141) as `_mm_blend_ps<const IMM4: i32>` and the body of the intrinsic will start with a static assertion ensuring IMM4 is indeed a 4-bit value. 

Note that I've linked to the more common macros, but this mechanism is applied throughout including arch-specific intrinsics and validations: e.g. [SAE/scale on Intel](https://github.com/rust-lang/stdarch/blob/c14e98417feb406df66e821ccd81e1293b4baa6f/crates/core_arch/src/x86/macros.rs), or [signed widths on MIPS](https://github.com/rust-lang/stdarch/blob/c14e98417feb406df66e821ccd81e1293b4baa6f/crates/core_arch/src/mips/msa/macros.rs).

Any intrinsic using these macros can cause an error like in the OP with an argument that doesn't pass the checks, but this set of functions should be equivalent to: all the intrinsics which have been ported to `#[rustc_legacy_const_generics]`.

**Comment by Mark-Simulacrum** [comment]
And the move to 'proper' const generics and the new validation are coupled because we would otherwise have undefined behavior (or be unable to codegen? -- what would happen)? Or can we decouple them?

**Comment by Amanieu** [comment]
We get LLVM assertion failures if invalid values reach codegen.

I'm not sure if decoupling is even possible. For example, these don't work in debug builds:
```rust
if VAL <= 3 {
    // The call to intrinsic is still emitted with an out-of-range value
    intrinsic(VAL);
} else {
    panic!();
}
```

```rust
if VAL <= 3 {
    // LLVM asserts because the argument is not a constant
    intrinsic(VAL & 3);
} else {
    panic!();
}
```

**Comment by RalfJung** [comment]
> I'm not sure if decoupling is even possible. 

It might be possible to do something involving specialization on const arguments... that seems rather fragile though.

An alternative might be to change the LLVM IR generated by rustc, but it seems like currently rustc has no idea about the structure of these LLVM intrinsics so that probably won't be easy either.

@Amanieu
> We get LLVM assertion failures if invalid values reach codegen.

How is it then possible that the example in the OP here compiles and runs on stable?

**Comment by Amanieu** [comment]
> How is it then possible that the example in the OP here compiles and runs on stable?

Previously stdarch had massive `match` statements to convert the value to a constant argument for the intrinsic.

**Comment by Mark-Simulacrum** [comment]
@craterbot retry

**Comment by Mark-Simulacrum** [comment]
@craterbot edit end=master#881c1ac408d93bb7adaa3a51dabab9266e82eee8 start=master#ff34b919075f35a1787659e9c448a34b06bab8de name=issue-85264

**Comment by Mark-Simulacrum** [comment]
@craterbot name=issue-85264 end=master#881c1ac408d93bb7adaa3a51dabab9266e82eee8 start=master#ff34b919075f35a1787659e9c448a34b06bab8de

**Comment by Mark-Simulacrum** [comment]
@craterbot retry

**Comment by Mark-Simulacrum** [comment]
@craterbot name=issue-85264 end=master#881c1ac408d93bb7adaa3a51dabab9266e82eee8 start=master#ff34b919075f35a1787659e9c448a34b06bab8de

**Comment by Amanieu** [comment]
All of the new failures that weren't in the previous report seem to come from https://github.com/ejmahler/RustFFT/issues/74, which is already fixed.

**Comment by lqd** [comment]
And it seems it's only 2 new crates that depend on this version of `rustfft`. So while post-monomorphization errors are tough errors to trigger and find consistently, the results are at least encouraging.

It can also be noted that the rough "evaluation of constant failed" error in the OP, has since improved on nightly, to somewhat acceptable levels.
<details>
<summary>Here's how it looks now</summary>

``` 
error[E0080]: evaluation of `core::core_arch::macros::ValidateConstImm::<51_i32, 0_i32, 15_i32>::VALID` failed
 --> /home/lqd/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/../../stdarch/crates/core_arch/src/macros.rs:8:9
  |
8 |         assert!(IMM >= MIN && IMM <= MAX, "IMM value not in expected range");
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'IMM value not in expected range', /rustc/dbe459dedd33470f2cb28101157de316caaffa66/library/core/src/../../stdarch/crates/core_arch/src/macros.rs:8:9
  |
  = note: this error originates in the macro `$crate::panic::panic_2015` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn std::arch::x86_64::_mm_blend_ps::<51_i32>`
    --> /home/lqd/.cargo/registry/src/github.com-1ecc6299db9ec823/rustfft-5.0.1/src/avx/avx_vector.rs:1314:23
     |
1314 |         let blended = _mm_blend_ps(rows[0], rows[2], 0x33);
     |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

</details>

And could be improved still if we, for example, specialized in `stdarch` the handful of actually used `ValidateConstImm` widths. It would allow mentioning the range bounds in the error message (or we could do some of the const fn string formatting hacks here of course).

**Comment by Amanieu** [comment]
Closing as working as intended. The error message has been improved, and a description of the potential breakage will be included in the 1.54 release notes (#86696).
